### ISSUE TITLE
Massive performance issue querying entities with many relations

### ISSUE THREAD

**Issue by m4tta** [typeorm/typeorm#7089]
## Issue Description
I have a query builder I am running to query for items in a table and loading their relations
```typescript
let deviceQuery = deviceRepo
    .createQueryBuilder('device')
    .leftJoinAndSelect('device.tankType', 'tankType')
    .leftJoinAndSelect('device.customer', 'customer')
    .leftJoinAndSelect('device.organization', 'organization')
    .leftJoinAndSelect('device.location', 'location')
  const devices: Device[] = await deviceQuery.getMany()
```
This table only has 7 items on it. When executing this query it takes upwards of 2500ms to return. If I replace `.getMany(`) with `.getRawMany()` the query only takes 140ms. I can also take the generated SQL from typeorm and paste it into PGAdmin and the query will also only take 140ms there. Something is going terribly wrong with the mapping process to the entity.

### Expected Behavior

For `.getMany()` to be performant when querying for entities that have _only_ 4 relations on them.

### Actual Behavior

`.getMany()` is 15x slower than `.getRawMany()`.

### Steps to Reproduce

1. Create an entity with 4+ `@ManyToOne()`/`@OneToMany()` relations
2. compare `.getMany()` to `.getRawMany()`

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |   Windows   |
| Node.js version     | v12.18.3 | 
| Typescript version  | v3.9.7 | 
| TypeORM version     | v0.2.29 | 

### Relevant Database Driver(s)

- [ ] `aurora-data-api`
- [ ] `aurora-data-api-pg`
- [ ] `better-sqlite3`
- [ ] `cockroachdb`
- [ ] `cordova`
- [ ] `expo`
- [ ] `mongodb`
- [ ] `mysql`
- [ ] `nativescript`
- [ ] `oracle`
- [x] `postgres`
- [ ] `react-native`
- [ ] `sap`
- [ ] `sqlite`
- [ ] `sqlite-abstract`
- [ ] `sqljs`
- [ ] `sqlserver`

### Are you willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time, and I know how to start.
- [ ] Yes, I have the time, but I don't know how to start. I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [x] No, I don't have the time and I wouldn't even know how to start.

**Comment by nebkat** [comment]
Could you create a minimal reproducible sample of this based on https://github.com/typeorm/typescript-example and I will try to tackle the problem. The results to entity mapping doesn't have anything that would immediately stand out as causing this problem so I'll have to debug it further.

Does the time scale linearly with each additional join, or exponentially?

**Comment by m4tta** [comment]
It's the end of my day here. I will try to get you the reproducible sample tomorrow.

As for the time scale

```typescript
  let deviceQuery = deviceRepo
    .createQueryBuilder('device') // 272ms
    .leftJoinAndSelect('device.tankType', 'tankType') // 441ms
    .leftJoinAndSelect('device.customer', 'customers') // 627ms
    .leftJoinAndSelect('device.organization', 'organization') // 1635ms
    .leftJoinAndSelect('device.location', 'location') // 2237ms
```
If I comment out all the leftJoins and add each one back in the order they appear these are the resulting total query times. Seems a bit exponential to me. If I run the SQL that gets printed when you do `.getSql()` in PgAdmin the query time is around sub ~200ms query times consistently. Same if I `.getRawMany()`.

**Comment by m4tta** [comment]
So I can't seem to replicate the issue in a minimal reproducible sample. It's turning out to be an odd issue at that. I am not even sure where the problem is anymore. I still think it's something to do with typeorm why would `getRawMany()` be 180ms, which is magnitudes faster than `getMany()` which takes about 2200ms to finish.

My initial DB setup is Amazon RDS with a m5.xlarge instance to rule out the instance size as the issue. Thinking maybe there was a problem with the RDS server. I loaded up a postgres docker image and ran postgres locally. imported a backup so both are working with the same small data set.

The locally ran postgres server performs flawlessly with typeorm. My queries now drop down to the 50-80ms range. This made me immediately think the problem is _obviously_ the AWS RDS server. But that still doesn't explain why the `getRawMany()` query runs fast, and `getMany()` runs slow. It's clear that typeorm is getting the data fast from the rawquery. It's just not mapping it fast. I'm just at a loss of what to do right now.

I can post all my entities if that would help.

**Comment by nebkat** [comment]
The difference between `getMany()` and `getRawMany()` is purely in TypeORM processing so it should happen regardless of the database you use. If it's not sensitive information to you it might be useful to see the entities - do they have an unusually large amount of columns or anything like that?

**Comment by m4tta** [comment]
Entities are available here https://gist.github.com/m4tta/9a86b260a2bd99b354dca946e5f93f2a

Largest one has about 25 columns. Most it is all basic types. I do use postgis geography type on 2 things.

**Comment by nebkat** [comment]
Have you checked with query logging if any additional queries are being run? The `@RelationId`s could be causing some slightly more expensive computation maybe?

**Comment by m4tta** [comment]
It was the `@RelationId`s that caused all the problems. When I would execute a `.getMany()` it would run about 11 queries. Picking one of these extra queries out and running it in pgadmin I noticed it was just giving me lists of Ids. This was a massive misunderstanding on my part on the cost of `@RelationId`. I thought they were free, and added paired it with every relationship I had even if it wasn't needed at the time.

Thank you so much for leading me to the right conclusion.

**Comment by nebkat** [comment]
Glad you got it figured out. `@RelationId` is still an experimental feature so hopefully that will be improved, but there should probably be more information about how it works in the background to avoid confusion like this. Generally the ORM hides the fact that one-to-many relations are much more complex than many-to-one. I personally avoid using them except for single item queries.

**Comment by imnotjames** [comment]
Duplicate of #3507

**Comment by dberardo-com** [comment]
i am experiencing a similar issue but i am not using @relationid ... also logging queries (loggin: true) does not give any additional hiddne query log.

i haven't tried using getmanyraw but i think i should because in my case i get a 10x slower query result in typeorm compared to the native query in the DB.

also the typeorm process and the DB are co-located on the same machine so network latency should be close to 0.

any hint how to go about this ?

**Comment by enioxt** [comment]
Hi\! Thanks for maintaining TypeORM.

**Observation:**
We reproduced slow queries when loading many relations via multiple LEFT JOINs (e.g., `.leftJoinAndSelect` across several relations). The query plan grows quickly and causes row explosion + heavy hydration.

**Proposed approach:**
- Prefer "query" relation load strategy or decouple loading:
  1) Fetch base entity IDs with a lean SELECT.
  2) Batch fetch related records using `where: { id: In(ids) }` per relation.
  3) Map in memory (lookup maps) to avoid join explosion.
- In 0.3+, `relationLoadStrategy: 'query'` often improves performance vs deep join trees.

**Expected impact (based on similar production fixes):**
- Fewer rows processed per query.
- Lower memory during hydration.
- Stable performance as relation count grows.

If helpful, I can provide a minimal repro with both patterns + timings and propose doc notes on "when to prefer queries over deep joins".

---
**Context:** We recently achieved 67x query reduction in a similar N+1 scenario (activepieces/activepieces#9744) using batch loading patterns.
